misc/pytorch_toolkit/noise_suppression/train.py

```diff
@@ -0,0 +1,568 @@
+            v = np.array(v)
+            if len(v.shape) == 1:
+                v = v[:, None]
+
+            if self.rank > -1:
+                # sync indicators
+                world_size = torch.distributed.get_world_size()
+                vt = torch.tensor(v).to(self.args.device)
+                torch.distributed.all_reduce(vt, op=torch.distributed.ReduceOp.SUM)
+                v = vt.cpu().numpy() / float(world_size)
+
+            self.indicators_mean[k] = v.mean(0)
+
+    def log_indicators(self, epoch_fp):
+        # Log metrics
+        str_out = ["ep {:.3f}".format(epoch_fp)]
+
+        lr = self.scheduler.get_last_lr()
+        str_out.append("lrp " + " ".join(['{:.2f}'.format(math.log10(t + EPS)) for t in lr]))
+
+        # add all indicators into log string
+        def tostr(v):
+            return " ".join(["{:.3f}".format(t) for t in v])
+        str_out.extend(["{} {}".format(k, tostr(v)) for k, v in self.indicators_mean.items()])
+
+        str_out.append("RC {}".format(self.check_loss_raise.reset_count))
+
+        # estimate processing times
+        if self.time_last is not None:
+            dt_iter = (time.time() - self.time_last) / len(self.indicators['loss'])
+            dt_ep = dt_iter * len(self.dataloader)
+            str_out.append("it {:.1f}s".format(dt_iter))
+            str_out.append("ep {:.1f}m".format(dt_ep / (60)))
+            str_out.append("eta {:.1f}h".format(dt_ep * (self.args.num_train_epochs - epoch_fp) / (60 * 60)))
+        self.time_last = time.time()
+
+        if self.rank in [-1, 0]:
+            logger.info(" ".join(str_out))
+
+    def save_and_eval_checkpoint(self, epoch):
+        if self.rank in [-1, 0]:
+            check_point_name = 'checkpoint-{:02}'.format(epoch)
+            output_dir = os.path.join(self.args.output_dir, check_point_name)
+            os.makedirs(output_dir, exist_ok=True)
+            self.model.save_pretrained(output_dir)
+            evaluate_dir(self.args.eval_data, output_dir)
+
+        if self.rank > -1:
+            torch.distributed.barrier()
+
+    def mix_signals(self, x_clean, x_noise):
+        # mix clean and noise signals on GPU
+        rms_noise = x_noise.pow(2).mean(-1, keepdim=True).sqrt()
+        rms_clean = x_clean.pow(2).mean(-1, keepdim=True).sqrt()
+
+        def rand_range(a, b):
+            return a + (b - a) * torch.rand(rms_clean.shape, dtype=rms_clean.dtype, device=rms_clean.device)
+
+        snr_db = rand_range(self.args.snr_min, self.args.snr_max)
+        target_db = rand_range(self.args.target_min, self.args.target_max)
+
+        def db_to_scale(db):
+            return 10 ** (db / 20)
+
+        # scale noise to get given SNR
+        scale_noise = db_to_scale(-snr_db) * rms_clean / (rms_noise + EPS)
+
+        # do not scale noise if noise only rms_clean < -80dB
+        #noise_only_w = (rms_clean < db_to_scale(-80)).float()
+        #scale_noise = 1 * noise_only_w + scale_noise * (1 - noise_only_w)
```

Comment:
Is this needed or it can be deleted? If this is needed, please, add condition when this code works instead of uncommenting it manually.

Reply:
deleted